core-backend/wasmi/src/funcs/mod.rs

```diff
@@ -1117,6 +1118,55 @@ where
         Func::wrap(store, func)
     }
 
+    pub fn random(store: &mut Store<HostState<E>>, forbidden: bool, memory: WasmiMemory) -> Func {
+        let func = move |mut caller: wasmi::Caller<'_, HostState<E>>,
+                         subject_ptr: u32,
+                         subject_len: u32,
+                         random_ptr: u32,
+                         bn_ptr: u32|
+              -> EmptyOutput {
+            exit_if!(forbidden, caller);
+
+            let mut subject = vec![0; subject_len as usize];
+
+            let read_result = {
+                let memory_wrap = get_caller_memory(&mut caller, &memory);
+                memory_wrap.read(subject_ptr as usize, subject.as_mut())
+            };
+
+            process_read_result!(read_result, caller);
+
+            subject.reserve(32);
+
+            let host_state = caller
+                .host_data()
+                .as_ref()
+                .expect("host_state should be set before execution");
```

Comment:
There is a macro: `let host_state = host_state_mut!(caller);`

Reply:
No need in mutable state